DESIGN.md

```diff
@@ -0,0 +1,818 @@
+is specified on the command line. The `-v` flag allows you to
+enable verbose logging.
+
+This command will emit two kind of messages. The first kind of
+messages consists of messages emitted by the measurement engine;
+the second kind consists of progress messages.
+
+Progress messages are like:
+
+```JSON
+{
+  "fields": {
+    "key": "performance.NDT",
+    "percentage": 0.7640000000000001,
+    "type": "progress"
+  },
+  "level": "info",
+  "timestamp": "2019-10-14T13:14:30.790923+02:00",
+  "message": "upload-speed (elapsed 9.55 s) 20996.89 kbit/s (num_streams 1)"
+}
+```
+
+where `key` identifies the experiment within the scope of the
+experiment group, `percentage` is the percentage of completion
+of the whole experiment group, and `message` is a message you
+may want to show to a user to further qualify what is happening.
+
+More specifically, `key` may be one of:
+
+- `"im.FacebookMessenger"`
+- `"im.Telegram"`
+- `"im.WhatsApp"`
+- `"middlebox.HTTPHeaderFieldManipulation"`
+- `"middlebox.HTTPInvalidRequestLine"`
+- `"performance.Dash"`
+- `"performance.NDT"`
+- `"websites.WebConnectivity"`
+
+Measurement engine messages are like:
+
+```JSON
+{
+  "fields": {
+    "type": "engine"
+  },
+  "level": "info",
+  "timestamp": "2019-10-03T16:52:56.327828+02:00",
+  "message": "starting facebook_messenger"
+}
+```
+
+They are warning, informational, or debugging messages emitted by
+the measurement engine to describe what it is doing.
+
+### Knowing the location of the home directory
+
+This command:
+
+```
+ooniprobe --batch info
+```
+
+emits this JSON:
+
+```JSON
+{
+  "fields": {
+    "path": "/home/sbs/.ooni"
+  },
+  "level": "info",
+  "timestamp": "2019-10-14T13:25:35.145144+02:00",
+  "message": "Home"
+}
+```
+
+where `message` is `"Home"` and `fields.path` contains the path to the
+currently used OONI's home directory. Of course, the output changes
+depending on the configured `OONI_HOME` environment variable value.
+
+## State and configuration directory
+
+This is `$HOME/.ooni` in all systems. The `ooniprobe --batch info`
+command can tell you the precise path on your system, subject to the
+override caused by the `OONI_HOME` environment variable.
+
+The state and configuration directory contains:
+
+- `assets`: directory where assets like GeoIP files are downloaded
+
+- `config.json`: default config file, used unless you specify
+another file using `--config <file>`
+
+- `db`: directory containing the SQLite3 database
+
+- `msmts`: directory containing the JSON measurements
+
+Running concurrent instances of the OONI CLI is unsupported. They
+may conflict attempting to download/update assets. This is the main
+reason why we recommend using another `OONI_HOME` when running the
+CLI from another program/app. We do this when running the CLI from
+OONI Desktop, and you should do the same in your use case.
+
+## Configuration file
+
+The configuration file has the following structure:
+
+```JSON
+{
```

Comment:
Some fields are missing here, namely:

```
  "_version": 0,
  "_informed_consent": true,
  "_is_beta": true,
  "auto_update": true,
```

Reply:
These are internal fields. I am not sure we want to document them. Probably not. I am also confused by the `auto_update` field. This is a setting for probe-desktop, as far as I get the design, so I believe it should not really be the probe-cli concern.

Reply:
> These are internal fields. I am not sure we want to document them

I guess so, though the desktop app needs to be aware of the `_informed_consent` field in order to know if onboarding should happen or not.

It's not 100% clear to me then what is the goal of this design document.

> I am also confused by the auto_update field. This is a setting for probe-desktop, as far as I get the design, so I believe it should not really be the probe-cli concern.

Yes you are right this is only used by the desktop app, though the settings are shared by both probe-cli and probe-desktop as in they both use the same configuration file.

Reply:
> It's not 100% clear to me then what is the goal of this design document.

The goal is that, if I give this document to a person, they could write a replacement for OONI Probe CLI that the OONI Probe Desktop could use via the `--batch` interface.

> Yes you are right this is only used by the desktop app, though the settings are shared by both probe-cli and probe-desktop as in they both use the same configuration file.
>
> [..]
>
> I guess so, though the desktop app needs to be aware of the _informed_consent field in order to know if onboarding should happen or not.

I was not aware of this design where the two tools share the same config file. I think I can see the advantages and the drawbacks of this design. I have not a good feeling about this design, and I'd like to spend more time thinking about whether it could serve us well or not.

(Returning to the original topic of the goal of writing a design document, I'd like to add that this is also an opportunity for further discussing and writing down existing assumptions.)